### PR TITLE
enable view changes test for pbft

### PR DESCRIPTION
**Pull request by olgavrou**
- client proxy clears reply certificates when there is a view change (all other certificates are cleared in replica during view change)
- setting the correct view on the encryptor during view changes
- handling prepares correctly during playback
- added `getCommitLocal` rpc for test to monitor the late joiners commit status without going through the protocol

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -540,7 +540,6 @@ if(BUILD_TESTS)
     NAME late_joiners
```
Comment (jumaffre) [inline CMakeLists.txt:540]: Could we rename this test to `late_joiners_pbft` as it is the only e2e test that is PBFT specific?

File: src/consensus/pbft/libbyz/client_proxy.h [hunk 2]
```diff
@@ -346,6 +348,15 @@ void ClientProxy<T, C>::recv_reply(Reply* reply)
             << " reply->full: " << (reply->full() ? "true" : "false")
             << " reps.cvalue: " << (void*)ctx->t_reps.cvalue() << std::endl;
 
+  if (current_view < reply->view())
+  {
+    LOG_INFO_FMT(
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/client_proxy.h:353]: `LOG_DEBUG_FMT`?

File: src/consensus/pbft/libbyz/replica.cpp [hunk 3]
```diff
@@ -1165,6 +1166,7 @@ void Replica::send_prepare(Seqno seqno, std::optional<ByzInfo> byz_info)
 
 void Replica::send_commit(Seqno s, bool send_only_to_self)
 {
+  LOG_DEBUG_FMT("Sending commit for seqno: {}", s);
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/replica.cpp:1169]: Is this going to be very verbose?

File: tests/late_joiners.py [hunk 4]
```diff
@@ -198,11 +198,10 @@ def run(args):
                         if args.consensus == "pbft"
                         else args.raft_election_timeout / 1000
                     )
-
                     if node.node_id == cur_primary_id and args.consensus == "pbft":
-                        # if pbft suspend the primary for twice the elecetion timeout
+                        # if pbft suspend the primary for > 2 the elecetion timeout
```
Comment (jumaffre) [inline tests/late_joiners.py:202]: Typo: election

File: tests/late_joiners.py [hunk 5]
```diff
@@ -198,11 +198,10 @@ def run(args):
...
                         # in order to make sure view changes will be triggered
-                        et += et
+                        et += 1.5 * et
```
Comment (eddyashton) [inline tests/late_joiners.py:204]: Minor: Took me a while to parse this. Can we keep the comment in English "suspend the primary for more than twice the election timeout", and make this line `et = et * 2.5` so it more obviously matches the comment? And maybe rename the variable from `et` to `suspend_time` (we modify the suspend time in this case, the election timeout is a constant).

File: src/consensus/pbft/libbyz/replica.cpp (hunk 3 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/replica.cpp:1169]: I keep adding this every time I debug pbft. It is a good indicator of what state the replica is in if it does send a commit message but then it doesn't add the commit message to it's certificate, so then you don't see the next debug message in `Replica::send_commit` (https://github.com/microsoft/CCF/blob/master/src/consensus/pbft/libbyz/replica.cpp#L1188)

Comment (achamayou) [reply src/consensus/pbft/libbyz/replica.cpp:1169]: Things that are per-transaction or more often should be TRACE, this is presumably per-batch so DEBUG or TRACE should be fine.

File: tests/late_joiners.py (hunk 4 above)
Comment (achamayou) [inline tests/late_joiners.py:202]:
```suggestion
                        # if pbft suspend the primary for > 2 the election timeout
```

Comment (achamayou) [reply tests/late_joiners.py:202]: For typos, I suggest we use the suggestion feature in GitHub as much as possible. It's often quicker than typing "Typo:" :)

File: CMakeLists.txt (hunk 1 above)
Comment (ashamis) [reply CMakeLists.txt:540]: From my understanding this should not be PBFT specific but this catches the known RAFT bug?

Comment (olgavrou) [reply CMakeLists.txt:540]: yes exactly

File: src/consensus/pbft/libbyz/replica.cpp (hunk 3 above)
Comment (ashamis) [reply src/consensus/pbft/libbyz/replica.cpp:1169]: Please use trace

File: CMakeLists.txt (hunk 1 above)
Comment (olgavrou) [reply CMakeLists.txt:540]: I'll change it to `late_joiners_pbft` for now and we can add one for raft later

**Pull request merged**